google/cloud/firestore_v1/throttle.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2021 Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+from typing import Callable, NoReturn, Optional
+
+import ipdb
+
+
+default_initial_tokens: int = 500
+default_phase_length: int = 60 * 5  # 5 minutes
+microseconds_per_second: int = 1000000
+
+
+class Throttle:
+    """Implements 5/5/5 ramp-up via Token Bucket algorithm."""
+
+    def __init__(
+        self,
+        initial_tokens: Optional[int] = default_initial_tokens,
+        phase_length: Optional[int] = default_phase_length,
+        now_getter: Optional[Callable] = datetime.datetime.utcnow,
+    ):
+        # Tracks the volume of operations during the
+        self._operations_this_phase: int = 0
+
+        # Optional override for telling time. Useful for testing.
+        self._now = now_getter
```

Comment:
Is using `mock.patch(`datetime.datetime.utcnow`), either within a testcase or as a wrapper around it, not a clearer pattern?  We've removed a bunch of `_now` thingies from other code over the years.  E.g., [here](https://github.com/googleapis/python-storage/blob/2c87f2f1438150bd93c7d544d2ccd4fdf95998db/tests/unit/test__signing.py#L96-L108).

Reply:
Hmm, potentially. My fear was that, as this is an extremely clock-sensitive feature, all the mocking and re-mocking to set the clock as just-so would end up being more tedious over all. But, if the removal of fake `_now`-type objects has already been a standing effort, that could swing the balance.

Reply:
Made this change.